exporter/opentelemetry-exporter-otlp-proto-http/src/opentelemetry/exporter/otlp/proto/http/metric_exporter/__init__.py

```diff
@@ -219,27 +243,448 @@ def export(
+                    )
+
+                    if split_resp.ok:
+                        export_result = MetricExportResult.SUCCESS
+                    elif self._retryable(split_resp):
+                        _logger.warning(
+                            "Transient error %s encountered while exporting metric batch, retrying in %ss.",
+                            split_resp.reason,
+                            delay,
+                        )
+                        sleep(delay)
+                        continue
+                    else:
+                        _logger.error(
+                            "Failed to export batch code: %s, reason: %s",
+                            split_resp.status_code,
+                            split_resp.text,
+                        )
+                        export_result = MetricExportResult.FAILURE
+
+                # Return result after all batches are attempted
+                return export_result
+
         return MetricExportResult.FAILURE
 
+    def _split_metrics_data(
+        self,
+        metrics_data: pb2.MetricsData,
+    ) -> Iterable[pb2.MetricsData]:
+        """Splits metrics data into several MetricsData (copies protobuf originals),
+        based on configured data point max export batch size.
+
+        Args:
+            metrics_data: metrics object based on HTTP protocol buffer definition
+
+        Returns:
+            Iterable[pb2.MetricsData]: An iterable of pb2.MetricsData objects containing
+                pb2.ResourceMetrics, pb2.ScopeMetrics, pb2.Metrics, and data points
+        """
+        batch_size: int = 0
+        # Stores split metrics data as editable references
+        # used to write batched pb2 objects for export when finalized
+        split_resource_metrics = []
+
+        for resource_metrics in metrics_data.resource_metrics:
+            split_scope_metrics = []
+            split_resource_metrics.append(
+                {
+                    "resource": resource_metrics.resource,
+                    "schema_url": resource_metrics.schema_url,
+                    "scope_metrics": split_scope_metrics,
+                }
+            )
+
+            for scope_metrics in resource_metrics.scope_metrics:
+                split_metrics = []
+                split_scope_metrics.append(
+                    {
+                        "scope": scope_metrics.scope,
+                        "schema_url": scope_metrics.schema_url,
+                        "metrics": split_metrics,
+                    }
+                )
+
+                for metric in scope_metrics.metrics:
+                    split_data_points = []
+
+                    # protobuf specifies metrics types (e.g. Sum, Histogram)
+                    # with different accessors for data points, etc
+                    # We maintain these structures throughout batch calculation
+                    current_data_points = []
+                    if metric.HasField("sum"):
```

Comment:
Another difference between http protobuf and the gRPC classes are that protobuf does not define a `DataPointT` to encompass all metric types. So I've added some `if-elif` throughout. Not sure if there is a better way.

Reply:
Unfortunate if this is the only way :(

Reply:
Yes, a bit easier to look at. Changed in [acc1c7c](https://github.com/open-telemetry/opentelemetry-python/pull/4576/commits/acc1c7c6f5746a401253f8f3a8af8b85f25829e3)